Wikipedia/Code/WMFReferencePopoverMessageViewController.m

```diff
@@ -0,0 +1,122 @@
+#import "Wikipedia-Swift.h"
+
+NSString *const WMFReferencePopoverShowNextNotification = @"WMFReferencePopoverShowNextNotification";
+NSString *const WMFReferencePopoverShowPreviousNotification = @"WMFReferencePopoverShowPreviousNotification";
+
+@interface WMFReferencePopoverMessageViewController ()<UITextViewDelegate>
+
+@property (strong, nonatomic) IBOutlet UITextView *textView;
+@property (strong, nonatomic) IBOutlet NSLayoutConstraint *widthConstraint;
+@property (strong, nonatomic) IBOutlet NSLayoutConstraint *horizontalSeparatorHeightConstraint;
+@property (strong, nonatomic) IBOutlet UIButton *closeButton;
+@property (strong, nonatomic) IBOutlet UILabel *titleLabel;
+
+@end
+
+@implementation WMFReferencePopoverMessageViewController
+
+-(void)setScrollEnabled:(BOOL)scrollEnabled {
+    self.textView.scrollEnabled = scrollEnabled;
+}
+
+-(BOOL)scrollEnabled {
+    return self.textView.scrollEnabled;
+}
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+
+    NSAssert(self.textView.scrollEnabled == NO, @"scrollEnabled must be NO for 'preferredContentSize' calculations to correctly account for textView's contentSize height");
+    
+    [self.widthConstraint setConstant:self.width];
+
+    self.textView.linkTextAttributes = @{NSForegroundColorAttributeName:[UIColor wmf_referencePopoverLinkColor]};
+    
+    [self.textView setAttributedText:[self attributedStringForHTML:[self referenceHTMLWithSurroundingHTML]]];
+    
+    self.horizontalSeparatorHeightConstraint.constant = 1.f / [UIScreen mainScreen].scale;
+    
+    self.closeButton.tintColor = [UIColor wmf_lightGrayColor];
+    
+    self.titleLabel.text = [[MWLocalizedString(@"reference-title", nil) uppercaseStringWithLocale:[NSLocale currentLocale]] stringByReplacingOccurrencesOfString:@"$1" withString:_linkText];
+}
+
+- (NSString*)referenceHTMLWithSurroundingHTML {
+    NSNumber *fontSize = [[NSUserDefaults wmf_userDefaults] wmf_readingFontSize];
+    
+    NSString *domain = [SessionSingleton sharedInstance].currentArticleSiteURL.wmf_language;
+    MWLanguageInfo *languageInfo = [MWLanguageInfo languageInfoForCode:domain];
+    NSString *baseUrl = [NSString stringWithFormat:@"https://%@.wikipedia.org/", languageInfo.code];
+
+    return
+    [NSString stringWithFormat:@""
+      "<html>"
+      "<head>"
+      "<base href='%@' target='_self'>"
+      "<style>"
+      " *{"
+      "     font-family:'-apple-system';"
+      "     font-size:16px;"
+      "     -webkit-text-size-adjust:%ld%%;"
+      "     color:#%@;"
+      "     text-decoration:none;"
+      " }"
+      "</style>"
+      "</head>"
+      "<body>"
+      "%@"
+      "</body>"
+      "</html>", baseUrl, (long)fontSize.integerValue, [[UIColor wmf_referencePopoverTextColor] wmf_hexStringIncludingAlpha:NO], self.HTML];
+}
+
+- (NSAttributedString*)attributedStringForHTML:(NSString*)html {
+    return
+    [[NSAttributedString alloc] initWithData:[html dataUsingEncoding:NSUnicodeStringEncoding]
```

Comment:
there used to be performance issues with this - has everything been tested on device?